kyuubi-common/src/main/scala/org/apache/kyuubi/engine/deploy/yarn/EngineYarnModeSubmitter.scala

```diff
@@ -0,0 +1,430 @@
+      val file = new File(path)
+      val files = Utils.listFilesRecursively(file)
+      files.foreach { f =>
+        if (f.isFile && f.canRead) {
+          distinctFiles += f
+        }
+      }
+    }
+    distinctFiles.groupBy(_.getName).map {
+      case (_, items) => items.head
+    }.toSeq
+  }
+
+  private def distribute(
+      srcPath: Path,
+      resType: LocalResourceType,
+      destName: String,
+      localResources: mutable.HashMap[String, LocalResource]): Unit = {
+    val fs = stagingDirPath.getFileSystem(hadoopConf)
+    val destPath = new Path(stagingDirPath, srcPath.getName)
+    info(s"Copying $srcPath to $destPath")
+    fs.copyFromLocalFile(srcPath, destPath)
+    fs.setPermission(destPath, new FsPermission(STAGING_DIR_PERMISSION))
+
+    val destFs = FileSystem.get(destPath.toUri, hadoopConf)
+    val destStatus = destFs.getFileStatus(destPath)
+
+    val destResource = Records.newRecord(classOf[LocalResource])
+    destResource.setType(resType)
+    destResource.setVisibility(LocalResourceVisibility.APPLICATION)
+    destResource.setResource(URL.fromPath(destPath))
+    destResource.setTimestamp(destStatus.getModificationTime)
+    destResource.setSize(destStatus.getLen)
+    localResources(destName) = destResource
+  }
+
+  private[kyuubi] def setupLaunchEnv(kyuubiConf: KyuubiConf): mutable.HashMap[String, String] = {
+    info("Setting up the launch environment for engine AM container")
+    val env = new mutable.HashMap[String, String]()
+
+    kyuubiConf.getAll
+      .filter { case (k, _) => k.startsWith(KyuubiConf.KYUUBI_ENGINE_YARN_MODE_ENV_PREFIX) }
+      .map { case (k, v) =>
+        (k.substring(KyuubiConf.KYUUBI_ENGINE_YARN_MODE_ENV_PREFIX.length + 1), v)
+      }
+      .foreach { case (k, v) => KyuubiHadoopUtils.addPathToEnvironment(env, k, v) }
+
+    addClasspathEntry(buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR), env)
+    env.put(
+      Environment.HADOOP_CONF_DIR.name(),
+      buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR, HADOOP_CONF_DIR))
+    addClasspathEntry(buildPath(Environment.PWD.$$(), LOCALIZED_CONF_DIR, HADOOP_CONF_DIR), env)
+    env.put("KYUUBI_ENGINE_YARN_MODE_STAGING_DIR", stagingDirPath.toString)
+    env
+  }
+
+  private def createApplicationSubmissionContext(
+      newApp: YarnClientApplication,
+      containerContext: ContainerLaunchContext): ApplicationSubmissionContext = {
+
+    val appContext = newApp.getApplicationSubmissionContext
+    appContext.setApplicationName(kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_APP_NAME)
+      .getOrElse(s"Apache Kyuubi $engineType Engine"))
+    appContext.setQueue(kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_QUEUE))
+    appContext.setAMContainerSpec(containerContext)
+    kyuubiConf.get(ENGINE_DEPLOY_YARN_MODE_PRIORITY).foreach { appPriority =>
+      appContext.setPriority(Priority.newInstance(appPriority))
+    }
+    appContext.setApplicationType(s"KYUUBI-${engineType.toUpperCase(Locale.ROOT)}-ENGINE")
```

Comment:
have discussed with @yaooqinn, let's make it configurable, and use `engineType.toUpperCase(Locale.ROOT)` by default, to align with spark

Reply:
we can hardcode `engineType.toUpperCase(Locale.ROOT)` in this PR and make it configurable later